### PR TITLE
Fix memory leak of EventDispatcher.Retry classes

### PR DESCRIPTION
**Pull request by sternr**
Using BlueOcean we noticed over time the amount of EventDispatcher.Retry classes in memory keeps climbing and are never GCd.
Looking at the code this is seems plausible in two different scenarios:

1. Unbounded retries on a faulty Retry object
2. Adding retry messages when the retry cleaner thread is not active
So the solution I proposed (looking good so far on our production environment) to add retry counter and make sure each time a retry object is added we add a timer task

See [JENKINS-51057](https://issues.jenkins-ci.org/browse/JENKINS-51057).

# Submitter checklist
- [x] Link to JIRA ticket in description, if appropriate.
- [ ] Change is code complete and matches issue description
- [ ] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

### CODE REVIEW

**Comment by olamy** [comment]
@sternr ping :)

**Comment by danielBelenky** [comment]
I think that the value of MAX_RETRY_COUNT should be configurable by the users.
Also, I think we should have the ability to limit the size of `retryQueue` and the value should
be configurable by users as well.

@sternr @olamy can I take over this PR?

**Comment by jonsten** [comment]
This PR is definitely a step in the right direction, unfortunately it still leads to a memory leak since old EventDispatcher objects are left behind and keep listen for events, even though the session/window has been terminated on the client side. In order to fix this properly there need to be some kind of mechanism om the server side which detects that the EventDispatcher has been abandoned by the client and should be unsubscribed.

As it is right now, including this fix, each session, e.g. window in Blue Ocean, will permanently allocate one EventDispatcher on the master. Sure it will not go as rouge as before the fix, but on long running instances with many users there will be a lot of EventDispatchers with a full queue, constantly listening for new messages and throwing away old.

So in short:
- Add some form of timeout mechanism to the EventDispatcher, if it hasn't been read for a while, it will unsubscribe and terminate
- Add limit to the number of EventDispatcher a session can have.

**Comment by sternr** [comment]
@jonsten we found the issue and are using it for the last few month successfully, I've now published it to our public repo:
https://github.com/jenkinsci/sse-gateway-plugin

**Comment by danielBelenky** [comment]
> @jonsten we found the issue and are using it for the last few month successfully, I've now published it to our public repo:

Nice! are you going to do a PR?

**Comment by sternr** [comment]
Yeah I'll make sure there are no new commits I missed and will submit the PR

**Comment by olamy** [comment]
@sternr regarding configuration look at the comments from my review. It can be as simple as system property.

**Comment by danielBelenky** [comment]
> old EventDispatcher objects are left behind and keep listen for events, even though the session/window has been terminated on the client side.

@jonsten can you elaborate on how it can happen? I see that each `HttpSession` a reference to it's associated `EventDispatcher`s so when a session times out/closes, the reference to the `EventDispatcher` no longer exists, hence this `EventDispatcher` can be GCd.

Or have I got it completely wrong?

**Comment by jonsten** [comment]
> > old EventDispatcher objects are left behind and keep listen for events, even though the session/window has been terminated on the client side.
> 
> @jonsten can you elaborate on how it can happen? I see that each `HttpSession` a reference to it's associated `EventDispatcher`s so when a session times out/closes, the reference to the `EventDispatcher` no longer exists, hence this `EventDispatcher` can be GCd.
> 
> Or have I got it completely wrong?

That was my understanding as well, but after looking at various heap dumps it is clear to me that was not the case. I.e. the out of memory bug fixed by this issue only occurs after several weeks and months of uptime where loads of EventDispatchers have been collecting events (which are never delivered). If there was proper timeout of the sessions these EventDispatchers would have been GCd and the memory leak would be under control. But for some reason they are never cleared from the sessions. As I mentioned in [this](https://issues.jenkins-ci.org/browse/JENKINS-51057?focusedCommentId=347743&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-347743) comment I'm able to find all the "leaked" instances in the sessions found by the following groovy code, `Jenkins.instance.servletContext.this$0._sessionHandler._sessionCache._sessions`.

Unfortunately I know too little about the session handling to determine if this is a separate issue, by design or related to this issue, but in either case, in my opinion, it is definitely the root cause to this memory leak. I.e. fix the session handling/caching of EventDispatchers and you nolonger have a problem with this bug. Sure it is still a small leak, but as soon as the session is terminated the EventDispatcher and its Retry objects will be GCd.

**Comment by danielBelenky** [comment]
@jonsten ,

In the Jira ticket you've mentioned, it seems like the problem is indeed `Retry` objects and not `EventDispatcher`.

Can you post a heap dump with leaked `EventDispatcher`s? In my instance, I don't see any leak in `EventDispatcher` objects. Only `Retry`.

Have you tried to reduce the default HTTP timeout in your server?

**Comment by jonsten** [comment]
@danielBelenky 
I must admit that it was a while since I was debugging this issue and I might be wrong on some points. But I will do my best do recap what we observed:
The problem that we were observing was that for our long running masters a significant portion of the heap was retained by the Retry instances and related objects. These Retry objects belonged to a relative few number of EventDispatcher objects, hundreds, definitely less that thousand. Each EventDispatcher object originated from a blue ocean session that a user had started. So basically we have had a handful of users opening blue oceans a few hundred times. Some of the EventDispatcher objects had been around for a long time (months) and collected a huge number of Retry objects (several for each build).

I can't recall why the sessions hadn't timed out, might be that the users always had a Jenkins window open which kept the session alive, or if we are affected by another bug which holds references to the session, in either case the session holds references to all EventDispatchers, even those whom the user have closed the corresponding Blue Ocean window. So if you for some reason have long running sessions EventDispatchers are kept.

I will see if I can find an old heap dump and investigate further, or if I can reproduce it, only problem is that we have uninstalled blue ocean in production and I've yet to reproduce the problem with sessions not timing out in sandbox.

With that said, regardless I argue that EventDispatchers is the problem here, the Retry objects are constructed and kept by the dispatchers (i.e. it is an internal privat class). And event dispatchers are leaked due to the fact that even though the user closes the blue ocean window, the corresponding EventDispatcher lives on on the server side, never to be used again. And as long as the user keeps the session alive (or due to other memory bug), the dispatcher will be there, listening on events.

Sorry for the long reply and arguing against the current solution, but I feel like this is just a workaround.

**Comment by danielBelenky** [comment]
@jonsten thank you for your detailed reply.

> And event dispatchers are leaked due to the fact that even though the user closes the blue ocean window, the corresponding EventDispatcher lives on on the server side, never to be used again.

I suspect that it might be an issue with BlueOcean and not the sse-gateway, but it's just an assumption based on what I've read in the code of `sse-gateway-plugin`. I didn't read the code of `Blue Ocean`.

In any case, I think that bounding the amount of `Retry` objects a dispatcher can hold and adding max retries for the `Retry` and dropping the `retryQueue` when reaching this max is a good enough solution to start with. From there we can continue to monitor and see if there's an issue with the `EventDispatcher` objects too.

**Comment by airadier** [comment]
Thanks everyone for this investigation. We found the related JENKINS-51057 and this PR after a similar investigation and heap dumps, and found many EventDispatcher objects were kept in the heap and never GCed. In our case, our Jenkins instance is exhausting heap memory (10GB) in around 7-10days. We are running the workaround script from JENKINS-51057 but very little memory is freed, so we still have lots of EventDispatcher objects that are not being collected.

@sternr I can see that apart from the original PR there are further commits in https://github.com/taboola/sse-gateway-plugin. Do this include any additional fix or are the changes focused on the Retry objects? I cannot see any PR from that repository.

Thanks and regards.

**Comment by sternr** [comment]
@airadier yes, we've been using our port for the last few month and have seen no leakage at all.

**Comment by olamy** [comment]
maybe we can close it except if this brings more feature than #30 
@sternr wdyt?

**Comment by olamy** [comment]
no answer so closing it
